### PR TITLE
ref #71: Implement a Blocking/Synchronous Error Reporting Method

### PR DESCRIPTION
**Pull request by akornich**
we also will have to add an extra section of documentation:

Blocking vs Non-Blocking Use of the Notifier

As an instrumentation component, the Notifier is designed to have as little impact on the hosting system/application as possible. 
Normally, you would want to use asynchronous logging since it has virtually no instrumentational overhead on your application execution performance at runtime. It has "fire and forget" approach to logging. 
However, in some specific situations, for example, while logging right before exiting an application, you may want to use it synchronously so that the application does not quit before the logging completes.
That is why all the logging methods of this interface imply asynchronous/non-blocking implementation. However, the interface defines the AsBlockingLogger(TimeSpan timeout) method that returns a synchronous implementation of ILogger. This approach allows for easier code refactoring when switching between asynchronous and synchronous uses of the logger.

So, to summarize all the info above, the following call:

logger.Log(ErrorLevel.Error, "test message");

would perform async logging, while this one:

logger.AsBlockingLogger(TimeSpan.FromSeconds(1)).Log(ErrorLevel.Error, "test message");

would perform blocking/synchronous logging with a timeout of 1 second.
In case of the timeout all the blocking log methods throw System.TimeoutException instead of gracefully completing the call. So, you might want to make all the blocking log calls within try-catch block while catching System.TimeoutException specifically to handle a timeout case as you wish.

### CODE REVIEW

File: Rollbar/RollbarLogger.cs [hunk 1]
```diff
@@ -209,25 +219,55 @@ public ILogger Debug(object obj, IDictionary<string, object> custom = null)
...
-        private void Report(System.Exception e, ErrorLevel? level = ErrorLevel.Error, IDictionary<string, object> custom = null)
+        internal void Report(
+            System.Exception e, 
+            ErrorLevel? level = ErrorLevel.Error, 
+            IDictionary<string, object> custom = null, 
+            TimeSpan? timeout = null,
+            SemaphoreSlim signal = null
+            )
         {
             SendBodyAsync(new Body(e), level, custom);
```
Comment (rokob) [inline Rollbar/RollbarLogger.cs]: Shouldn't this pass the `timeout` and `signal` arguments?

File: Rollbar/RollbarQueueController.cs [hunk 2]
```diff
@@ -214,6 +219,14 @@ private void ProcessQueues(AccessTokenQueuesMetadata tokenMetadata)
             }
         }
 
+        private void ObeyPayloadTimeout(Payload payload, PayloadQueue queue)
+        {
+            if (payload.TimeoutAt.HasValue && (DateTime.Now.Add(this._sleepInterval) >= payload.TimeoutAt.Value))
+            {
+                queue.Dequeue();
```
Comment (rokob) [inline Rollbar/RollbarQueueController.cs]: Doesn't this mean that we would not signal the signal inside the payload?

File: Rollbar/RollbarLogger.cs [hunk 3]
```diff
@@ -209,25 +219,55 @@ public ILogger Debug(object obj, IDictionary<string, object> custom = null)
...
+            Body body, ErrorLevel? level, 
+            IDictionary<string, object> custom,
+            TimeSpan? timeout = null,
+            SemaphoreSlim signal = null
+            )
         {
+            DateTime? timeoutAt = null;
+            if (timeout.HasValue)
+            {
+                timeoutAt = DateTime.Now.Add(timeout.Value);
```
Comment (rokob) [inline Rollbar/RollbarLogger.cs]: What does the timeout mean if you set the `timeoutAt` time here? It seems like we use the `timeout` value a couple different ways.

File: UnitTest.Rollbar/RollbarLoggerBlockingWrapperFixture.cs [hunk 4]
```diff
@@ -0,0 +1,87 @@
+﻿#pragma warning disable CS1591 // Missing XML comment for publicly visible type or member
+
```
Comment (rokob) [inline UnitTest.Rollbar/RollbarLoggerBlockingWrapperFixture.cs]: Any tests that use actual time and timeouts are fragile and usually force tests to take longer than they need to. Testing timeout behaviour should be possible without having to actually wait for things to timeout or not. These tests will fail at some point when something takes longer to run that it usually does. We need a different way of injecting time to be able to test this better.

File: Rollbar/RollbarLogger.cs (hunk 1 above)
Comment (akornich) [reply Rollbar/RollbarLogger.cs]: absolutely - should. thanks! fixed it.

File: Rollbar/RollbarQueueController.cs (hunk 2 above)
Comment (akornich) [reply Rollbar/RollbarQueueController.cs]: we signal the semaphore only in case of successful transmission of the payload to the rollbar api service, otherwise it just times out and gets disposed of in either case at the end of the blocking method.

File: Rollbar/RollbarLogger.cs (hunk 3 above)
Comment (akornich) [reply Rollbar/RollbarLogger.cs]: "timeout" is a time span specifying for how long to wait for sending out a payload.
"timeoutAt" is a cut off time to give up sending the payload and to discard the payload. it is calculated at the time of allocating the payload plus the "timeout". based on this timestamp the queues processing thread knows when to give up transmission of any particular payload.

File: UnitTest.Rollbar/RollbarLoggerBlockingWrapperFixture.cs (hunk 4 above)
Comment (akornich) [reply UnitTest.Rollbar/RollbarLoggerBlockingWrapperFixture.cs]: agree. but here we are waiting for the expected timeout of a blocking call while also making sure that that it is not broken and the timeout exception is thrown. btw, we do have a few other unit test methods protected by a timeout attribute just in case to avoid "hanging" of the test in case the unit under test malfunctions. i normally set these guarding timeout as multiples of "tested timeout" somewhat unreasonably large for the tested scenario. so, if it does fail due to the delayed over time execution logic, we probably do want to look into it anyway.

**Comment by rokob** [comment]
LGTM
